### DISCUSSION TITLE
Action does not work when Store is created using initProps

### DISCUSSION THREAD

**Discussion by pdhruv93** [pmndrs/zustand#2189] [Bug report] [answered]
Here is the link to [sandbox](https://codesandbox.io/p/sandbox/react-zustand-init-d7c42y?file=%2Fsrc%2FNameStore.ts%3A1%2C1&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clopqf5ad00072e68f2gz2d1e%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clopqf5ad00032e689idoogwx%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clopqf5ad00052e6807zmu90b%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clopqf5ad00062e68dj1kmmml%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B60%252C40%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clopqf5ad00032e689idoogwx%2522%253A%257B%2522id%2522%253A%2522clopqf5ad00032e689idoogwx%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522id%2522%253A%2522clopqkxiq00pj2e68891q5ona%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FNameStore.ts%2522%252C%2522id%2522%253A%2522clopqlygf00ve2e68fos3exyj%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clopqlygf00ve2e68fos3exyj%2522%257D%252C%2522clopqf5ad00062e68dj1kmmml%2522%253A%257B%2522id%2522%253A%2522clopqf5ad00062e68dj1kmmml%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clopqkfqm00fw2e686zlr9x0k%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%253Fname%253Ddhruv%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clopqkfqm00fw2e686zlr9x0k%2522%257D%252C%2522clopqf5ad00052e6807zmu90b%2522%253A%257B%2522id%2522%253A%2522clopqf5ad00052e6807zmu90b%2522%252C%2522activeTabId%2522%253A%2522clopqf5ad00042e68b33217yb%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clopqf5ad00042e68b33217yb%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clopqf5p3000degefhe694ck3%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522id%2522%253A%2522clopqg121009d2e68meexr2nv%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)
The updateName action doesn't do anything. I think I am doing it wrong. I am using the createStore Hook to use the action.  But whats the correct way to implement it?

**Comment by dbritto-dev** [comment]
@pdhruv93 here you go -> https://codesandbox.io/p/sandbox/react-zustand-init-forked-nzp4qy?file=%2Fsrc%2FApp.tsx%3A15%2C11

**Reply by dbritto-dev** [reply]
Some things you need to know about `create` is that returns a custom hook, when you wrap `create` inside of other function you are creating a `custom hook factory` (in a nutshell a closure), so in order to create the custom hook you need to call your function that creates the custom hook and then use that custom hook. For instance

```ts
const createUseCounter = (initialCount = 0) => {
  return create((set) => ({
    count: initialCount,
    inc: () => set((state) => ({ count: state.count + 1 })),
  }));
}
```

```tsx
const Counter = () => {
  const useCounterRef = useRef()
  if (!useCounterRef.current) {
    useCounterRef.current = createUseCounter()
  }
  const { count, inc } = useCounterRef.current()

  return <button type="button" onClick={inc}>{count}</button>
}
```

**Reply by pdhruv93** [reply]
Thanks a lot @dbritto-dev. The action works now. The only challenge I have now is with the store subscription:
```
useNameStore().subscribe(
        (state) => [
            state.name
        ],
        () => doSomething(),
        { equalityFn: shallow }
    );
```

Getting error `Expected 1 arguments, but got 3.` [Here is the link to the sandbox for quick check](https://codesandbox.io/p/sandbox/react-zustand-init-fixed-756748?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clopvx33v00092e69e7srgtdd%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clopvx33v00042e69dminu2r6%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clopvx33v00072e69yckz1vjo%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clopvx33v00082e69mupzuf3t%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B60%252C40%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clopvx33v00042e69dminu2r6%2522%253A%257B%2522id%2522%253A%2522clopvx33v00042e69dminu2r6%2522%252C%2522activeTabId%2522%253A%2522clopvx33v00032e69noi9xq9e%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clopvx33v00032e69noi9xq9e%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FNameStore.ts%2522%252C%2522id%2522%253A%2522clopw4p3600a72e680tqbnggl%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%257D%252C%2522clopvx33v00082e69mupzuf3t%2522%253A%257B%2522id%2522%253A%2522clopvx33v00082e69mupzuf3t%2522%252C%2522activeTabId%2522%253A%2522clopw5pkl00h42e68tpfug1ur%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522SANDBOX_INFO%2522%252C%2522id%2522%253A%2522clopw2v5m00dt2e68ioxex6uk%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clopw5pkl00h42e68tpfug1ur%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clopvx33v00072e69yckz1vjo%2522%253A%257B%2522id%2522%253A%2522clopvx33v00072e69yckz1vjo%2522%252C%2522activeTabId%2522%253A%2522clopw5eyx00g02e68ettindxr%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clopqf5p3000degefhe694ck3%2522%252C%2522id%2522%253A%2522clopw5eyx00g02e68ettindxr%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A22.045779685264662%257D)

**Reply by dbritto-dev** [reply, marked as answer]
@pdhruv93 here you go -> https://codesandbox.io/p/sandbox/react-zustand-init-forked-nzp4qy?file=%2Fsrc%2FApp.tsx%3A26%2C7

**Reply by dbritto-dev** [reply]
@pdhruv93 keep me in the loop

**Reply by pdhruv93** [reply]
@dbritto-dev Works perfectly. A new problem now is that the state is not getting shared between components. It is creating a new store for each component. [Sandbox link](https://codesandbox.io/p/sandbox/react-zustand-init-fixed-756748?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clopvx33v00092e69e7srgtdd%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clopvx33v00042e69dminu2r6%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clopvx33v00072e69yckz1vjo%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clopvx33v00082e69mupzuf3t%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B60%252C40%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clopvx33v00042e69dminu2r6%2522%253A%257B%2522id%2522%253A%2522clopvx33v00042e69dminu2r6%2522%252C%2522activeTabId%2522%253A%2522clopvx33v00032e69noi9xq9e%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clopvx33v00032e69noi9xq9e%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FOtherComponent.tsx%2522%252C%2522id%2522%253A%2522clopze82500nu2e697xxr95oy%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FNameStore.ts%2522%252C%2522id%2522%253A%2522clopzj0fg03882e69mtg4slox%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%257D%252C%2522clopvx33v00082e69mupzuf3t%2522%253A%257B%2522id%2522%253A%2522clopvx33v00082e69mupzuf3t%2522%252C%2522activeTabId%2522%253A%2522clopw5pkl00h42e68tpfug1ur%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522SANDBOX_INFO%2522%252C%2522id%2522%253A%2522clopw2v5m00dt2e68ioxex6uk%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%252C%257B%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clopw5pkl00h42e68tpfug1ur%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clopvx33v00072e69yckz1vjo%2522%253A%257B%2522id%2522%253A%2522clopvx33v00072e69yckz1vjo%2522%252C%2522activeTabId%2522%253A%2522clopw5eyx00g02e68ettindxr%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clopqf5p3000degefhe694ck3%2522%252C%2522id%2522%253A%2522clopw5eyx00g02e68ettindxr%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A22.045779685264662%257D).

**Reply by dbritto-dev** [reply]
@pdhruv93 yeah, that's expected if you want to share store then you need to use `useContext` and `Context API` , something like this -> https://codesandbox.io/p/sandbox/gifted-dan-m2f8zj?file=%2Fsrc%2FApp.tsx%3A29%2C32

**Reply by pdhruv93** [reply]
Aha. The [zustand docs](https://docs.pmnd.rs/zustand/guides/initialize-state-with-props) starts making sense now after this. All of this is needed for creating store from init props. Thanks a lot for such quick response!

**Reply by LeoCaprile** [reply]
@dbritto-dev, hey I actually used `useMemo` instead of `useRef,` are any drawbacks on using `useMemo` insted of `useRef`? I mean I undestand that using `useRef` is the "docs way" but using `useMemo` made more sense to me and a seems more cleaner solution. What do you think?

**Reply by dbritto-dev** [reply]
Actually, the usage of useMemo doesn't make more sense here. Let me take a piece of the useMemo react caveats as a reference:

> React will not throw away the cached value unless there is a specific reason to do that. For example, in development, React throws away the cache when you edit the file of your component. Both in development and in production, React will throw away the cache if your component suspends during the initial mount. In the future, React may add more features that take advantage of throwing away the cache—for example, if React adds built-in support for virtualized lists in the future, it would make sense to throw away the cache for items that scroll out of the virtualized table viewport. This should be fine if you rely on useMemo solely as a performance optimization. Otherwise, a state variable or a ref may be more appropriate.

In a nutshell you should only use useMemo for performance optimization not for initialization

**Reply by LeoCaprile** [reply]
> Actually, the usage of useMemo doesn't make more sense here. Let me take a piece of the useMemo react caveats as a reference:
> 
> > React will not throw away the cached value unless there is a specific reason to do that. For example, in development, React throws away the cache when you edit the file of your component. Both in development and in production, React will throw away the cache if your component suspends during the initial mount. In the future, React may add more features that take advantage of throwing away the cache—for example, if React adds built-in support for virtualized lists in the future, it would make sense to throw away the cache for items that scroll out of the virtualized table viewport. This should be fine if you rely on useMemo solely as a performance optimization. Otherwise, a state variable or a ref may be more appropriate.
> 
> In a nutshell you should only use useMemo for performance optimization not for initialization

I understand now, thanks for the reply! makes more sense now, anyways, I liked the way it looks with the useMemo hahah, but I'm going to use `useRef`.

**Comment by pdhruv93** [comment]
@dbritto-dev Isn't there some other way to populate the store(also share the store) other than using the context? I kind of feel that using context is killing the flexibility of zustand store. I can only access the store and store state under the context provider.

One way I find is to just to create an action under store to set the states to input props. But setting the states this way is causing the subscription to run

**Reply by dbritto-dev** [reply]
@pdhruv93 it depends on your needs if you are building an SPA application you normally don't need to use context but talking about nextjs that's a requirement and also a recommendation

We use to use `createStore` instead of `create` since that's the recommended way

**Reply by pdhruv93** [reply]
@dbritto-dev I have this simple [sandbox](https://codesandbox.io/p/devbox/pensive-water-756748?file=%2Fsrc%2Fstore%2FNameStore.ts%3A9%2C3&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clt7f2sjf00063b6h8xf8azvb%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clt7f2sjf00023b6hid8gfe94%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clt7f2sjf00043b6hkkw0bvbt%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clt7f2sjf00053b6hbjsftehb%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B69.59221935786266%252C30.40778064213734%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clt7f2sjf00023b6hid8gfe94%2522%253A%257B%2522id%2522%253A%2522clt7f2sjf00023b6hid8gfe94%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clt7f2sjf00053b6hbjsftehb%2522%253A%257B%2522id%2522%253A%2522clt7f2sjf00053b6hbjsftehb%2522%252C%2522activeTabId%2522%253A%2522clt7f3sg700av3b6hbpunbtfh%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clt7f3sg700av3b6hbpunbtfh%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clt7f2sjf00043b6hkkw0bvbt%2522%253A%257B%2522id%2522%253A%2522clt7f2sjf00043b6hkkw0bvbt%2522%252C%2522activeTabId%2522%253A%2522clt7f3o2200973b6h8ep22xy3%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clt7f2sjf00033b6h128j7hyc%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clt7f2tf1000ydka75aaoga1p%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522id%2522%253A%2522clt7f3o2200973b6h8ep22xy3%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) using Context and createStore

Changing fName also re-renders the other component where I am not using fName from store. Is there something wrong or it is as expected? If that's the case what's the point to Zustand to highlight render optimization and the changes impact only the components that are subscribed?

**Reply by dbritto-dev** [reply]
@pdhruv93 the right way to avoid extra renders is by using `useStoreWithEqualityFn` from `zustand/traditional` instead of `useStore` from `zustand` so you hook should look like this:

```ts
export function useNameStoreContext<T>(selector: (state: NameStore) => T): T {
  const context = useContext(NameStoreContext);
  if (!context) {
    throw new Error("useNameStore must be used within NameStoreProvider");
  }
  return useStoreWithEqualityFn(context, selector, shallow);
}
```

> **Note:** for more complex cases you want to use a deep equality fn instead of `shallow`

**Reply by pdhruv93** [reply]
Thanks a lot @dbritto-dev , you have been super helpful all the time

**Comment by pdhruv93** [comment]
@dbritto-dev Is it possible to access the current store params from within the subscription. I would like to make a call to the api sending in all the store params. I could only get the current state of the subscribed store params

```
nameStore.subscribe(
        (state) => [state.fName, state.lName],
        () => {GET ALL THE STORE PARAMS HERE, NOT JUST WHAT ARE SUBSCRIBED},
        { equalityFn: shallow },
      );
```

**Reply by dbritto-dev** [reply]
> @dbritto-dev Is it possible to access the current store params from within the subscription. I would like to make a call to the api sending in all the store params. I could only get the current state of the subscribed store params
> 
> 
> 
> ```
> 
> nameStore.subscribe(
> 
>         (state) => [state.fName, state.lName],
> 
>         () => {GET ALL THE STORE PARAMS HERE, NOT JUST WHAT ARE SUBSCRIBED},
> 
>         { equalityFn: shallow },
> 
>       );
> 
> ```

@pdhruv93 yes, it's possible with `nameStore.getState()`

**Comment by pdhruv93** [comment]
@dbritto-dev everything now works, the store is also populated. One problem:
The store is getting recreated multiple times, the first time with the correct initProps and after that the initProps are getting erased:

I logged this:
```
export const useNameStore = (initProps?: Partial<NameStoreProps>) => {
    const useNameStoreRef = useRef<ReturnType<typeof createNameStore>>(
        undefined as never
    );
    console.log(':::Creating store');

    if (!useNameStoreRef.current) {
        console.log(':::Creating new', initProps);
        useNameStoreRef.current = createNameStore(initProps);
    }
    return useNameStoreRef.current;
};
```

I see following logs:
```
:::Creating store
:::Creating new {fName: 'fff', lName: 'lll'}
:::Creating store
:::Creating store
:::Creating store
```

**Reply by dbritto-dev** [reply]
@pdhruv93 actually, only creating new log would be considered in this case; so creating store logs are fine

**Reply by pdhruv93** [reply]
Then what might be the reason my store is getting cleared? When I log some state from store, it prints undefined, only first time it is printed.

**Reply by pdhruv93** [reply]
aah it is something else related to my other code, store works fine, thanks!

**Reply by dbritto-dev** [reply]
@pdhruv93 hmmm would you mind sharing us a minimal repro on stackblitz? btw, we recommend use `createStore` instead of `create` when you want to use a `custom hook factory`

**Reply by pdhruv93** [reply]
Here it is on [codesandbox](https://codesandbox.io/p/devbox/pensive-water-756748?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clt7f2sjf00063b6h8xf8azvb%2522%252C%2522sizes%2522%253A%255B100%252C0%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clt7f2sjf00023b6hid8gfe94%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clt7f2sjf00043b6hkkw0bvbt%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clt7f2sjf00053b6hbjsftehb%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B61.49175460951894%252C38.50824539048106%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clt7f2sjf00023b6hid8gfe94%2522%253A%257B%2522id%2522%253A%2522clt7f2sjf00023b6hid8gfe94%2522%252C%2522tabs%2522%253A%255B%255D%257D%252C%2522clt7f2sjf00053b6hbjsftehb%2522%253A%257B%2522id%2522%253A%2522clt7f2sjf00053b6hbjsftehb%2522%252C%2522activeTabId%2522%253A%2522clt7f3sg700av3b6hbpunbtfh%2522%252C%2522tabs%2522%253A%255B%257B%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522port%2522%253A3000%252C%2522id%2522%253A%2522clt7f3sg700av3b6hbpunbtfh%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522path%2522%253A%2522%252F%2522%257D%255D%257D%252C%2522clt7f2sjf00043b6hkkw0bvbt%2522%253A%257B%2522id%2522%253A%2522clt7f2sjf00043b6hkkw0bvbt%2522%252C%2522activeTabId%2522%253A%2522clt7f3o2200973b6h8ep22xy3%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clt7f2sjf00033b6h128j7hyc%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522clt7f2tf1000ydka75aaoga1p%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522start%2522%252C%2522id%2522%253A%2522clt7f3o2200973b6h8ep22xy3%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Afalse%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D). Let me know it anything else is needed too

**Reply by dbritto-dev** [reply]
@pdhruv93 everything looks good to me
